lightGBM 2.2.200 still has glibc issue /lib64/libstdc++.so.6: version : GLIBCXX_3.4.20' not found

Thank you @xiaoxiaofengzi for the clarification!
Unfortunately this is a known problem caused by that MMLSpark uses old and broken (in terms of dependencies) version of LightGBM.

You can track this [PR](https://github.com/Azure/mmlspark/pull/391) and these issues [1](https://github.com/Azure/mmlspark/issues/335), [2](https://github.com/Azure/mmlspark/issues/390) to know when the problem will be solved on the Spark-package side. I think that you can copy-paste your log there to hurry them gently.
From the our side starting from `2.2.2` version we automatically guarantee `GLIBC <= 2.14` and `GLIBCXX <= 3.4.19`:

https://github.com/Microsoft/LightGBM/blob/a694712b7fb86cd532eea2c1781b58d4ba58436a/helpers/check_dynamic_dependencies.py#L18-L23
https://github.com/Microsoft/LightGBM/blob/a694712b7fb86cd532eea2c1781b58d4ba58436a/helpers/check_dynamic_dependencies.py#L25-L31

Thanks for your patience and sorry for the inconvenience!

_Originally posted by @StrikerRUS in https://github.com/Microsoft/LightGBM/issues/1858#issuecomment-440610638_

lightGBM 2.2.2 still can not run in older system (centOS 7):
![image](https://user-images.githubusercontent.com/11178311/51102418-d9de8a00-1819-11e9-895d-1dd4cebc105c.png)
jar:exception
Caused by: java.lang.UnsatisfiedLinkError: /tmp/mml-natives1380894375751123942/lib_lightgbm.so: /lib64/libstdc++.so.6: version `GLIBCXX_3.4.20' not found (required by /tmp/mml-natives1380894375751123942/lib_lightgbm.so)
my environment:
Linux tw-node41 3.10.0-327.el7.x86_64 #1 SMP Thu Nov 19 22:10:57 UTC 2015 x86_64 x86_64 x86_64 GNU/Linux 
ldd (GNU libc) 2.17

=======================================

i could compile the lightGBM source in this environment successfully,
but get the same exception when i run the example in folders.
![2019-01-11 18-05-35](https://user-images.githubusercontent.com/11178311/51102913-af8dcc00-181b-11e9-904a-7c1c966ae5c2.png)

--- a/.ci/dockers/ubuntu-14.04/Dockerfile
+++ b/.ci/dockers/ubuntu-14.04/Dockerfile
@@ -41,6 +41,26 @@ RUN curl -sL https://cmake.org/files/v3.14/cmake-3.14.1-Linux-x86_64.sh -o cmake
  && ./cmake.sh --prefix=/usr/local --exclude-subdir \
  && rm cmake.sh
 
+# Install Java
+RUN add-apt-repository ppa:openjdk-r/ppa -y \
+ && apt-get update \
+ && apt-get install -y --no-install-recommends \
+    openjdk-8-jdk \
+ && rm -rf /var/lib/apt/lists/*
+
+ENV JAVA_HOME=/usr/lib/jvm/java-8-openjdk-amd64/
+
+# Install SWIG
+RUN curl -sL https://downloads.sourceforge.net/project/swig/swig/swig-3.0.12/swig-3.0.12.tar.gz -o swig.tar.gz \
+ && tar -xzf swig.tar.gz \
+ && cd swig-3.0.12 \
+ && ./configure --prefix=/usr/local --without-pcre \
+ && make \
+ && make install \
+ && cd .. \
+ && rm swig.tar.gz \
+ && rm -rf swig-3.0.12
+
 # Install Miniconda
 RUN curl -sL https://repo.continuum.io/miniconda/Miniconda3-latest-Linux-x86_64.sh -o miniconda.sh \
  && chmod +x miniconda.sh \

--- a/.ci/test.sh
+++ b/.ci/test.sh
@@ -74,9 +74,9 @@ fi
 
 if [[ $TASK == "if-else" ]]; then
     conda install -q -y -n $CONDA_ENV numpy
-    mkdir $BUILD_DIRECTORY/build && cd $BUILD_DIRECTORY/build && cmake "${CMAKE_OPTS[@]}" .. && make -j4 || exit -1
+    mkdir $BUILD_DIRECTORY/build && cd $BUILD_DIRECTORY/build && cmake "${CMAKE_OPTS[@]}" .. && make lightgbm -j4 || exit -1
     cd $BUILD_DIRECTORY/tests/cpp_test && ../../lightgbm config=train.conf convert_model_language=cpp convert_model=../../src/boosting/gbdt_prediction.cpp && ../../lightgbm config=predict.conf output_result=origin.pred || exit -1
-    cd $BUILD_DIRECTORY/build && make -j4 || exit -1
+    cd $BUILD_DIRECTORY/build && make lightgbm -j4 || exit -1
     cd $BUILD_DIRECTORY/tests/cpp_test && ../../lightgbm config=predict.conf output_result=ifelse.pred && python test.py || exit -1
     exit 0
 fi
@@ -137,7 +137,7 @@ else
     cmake "${CMAKE_OPTS[@]}" ..
 fi
 
-make -j4 || exit -1
+make _lightgbm -j4 || exit -1
 
 cd $BUILD_DIRECTORY/python-package && python setup.py install --precompile --user || exit -1
 pytest $BUILD_DIRECTORY/tests || exit -1

--- a/.ci/test.sh
+++ b/.ci/test.sh
@@ -18,6 +18,19 @@ else
     CMAKE_OPTS=()
 fi
 
+if [[ $AZURE == "true" ]] && [[ $OS_NAME == "linux" ]] && [[ $TASK == "swig" ]]; then
+    mkdir $BUILD_DIRECTORY/build && cd $BUILD_DIRECTORY/build
+    cmake -DUSE_SWIG=ON "${CMAKE_OPTS[@]}" ..
+    make -j4 || exit -1
+    if [[ $COMPILER == "gcc" ]]; then
+        objdump -T $BUILD_DIRECTORY/lib_lightgbm.so > $BUILD_DIRECTORY/objdump.log || exit -1
+        objdump -T $BUILD_DIRECTORY/lib_lightgbm_swig.so >> $BUILD_DIRECTORY/objdump.log || exit -1
+        python $BUILD_DIRECTORY/helpers/check_dynamic_dependencies.py $BUILD_DIRECTORY/objdump.log || exit -1
+    fi
+    cp $BUILD_DIRECTORY/build/lightgbmlib.jar $BUILD_ARTIFACTSTAGINGDIRECTORY/lightgbmlib.jar
+    exit 0
+fi
+
 conda create -q -y -n $CONDA_ENV python=$PYTHON_VERSION
 source activate $CONDA_ENV
 
@@ -61,9 +74,9 @@ fi
 
 if [[ $TASK == "if-else" ]]; then
     conda install -q -y -n $CONDA_ENV numpy
-    mkdir $BUILD_DIRECTORY/build && cd $BUILD_DIRECTORY/build && cmake "${CMAKE_OPTS[@]}" .. && make lightgbm -j4 || exit -1
+    mkdir $BUILD_DIRECTORY/build && cd $BUILD_DIRECTORY/build && cmake "${CMAKE_OPTS[@]}" .. && make -j4 || exit -1
     cd $BUILD_DIRECTORY/tests/cpp_test && ../../lightgbm config=train.conf convert_model_language=cpp convert_model=../../src/boosting/gbdt_prediction.cpp && ../../lightgbm config=predict.conf output_result=origin.pred || exit -1
-    cd $BUILD_DIRECTORY/build && make lightgbm -j4 || exit -1
+    cd $BUILD_DIRECTORY/build && make -j4 || exit -1
     cd $BUILD_DIRECTORY/tests/cpp_test && ../../lightgbm config=predict.conf output_result=ifelse.pred && python test.py || exit -1
     exit 0
 fi
@@ -124,7 +137,7 @@ else
     cmake "${CMAKE_OPTS[@]}" ..
 fi
 
-make _lightgbm -j4 || exit -1
+make -j4 || exit -1
 
 cd $BUILD_DIRECTORY/python-package && python setup.py install --precompile --user || exit -1
 pytest $BUILD_DIRECTORY/tests || exit -1

--- a/.vsts-ci.yml
+++ b/.vsts-ci.yml
@@ -10,7 +10,7 @@ jobs:
 - job: Linux
 ###########################################
   variables:
-    COMPILER: gcc
+    COMPILER: clang
   pool:
     vmImage: 'ubuntu-16.04'
   container: ubuntu1404

--- a/.vsts-ci.yml
+++ b/.vsts-ci.yml
@@ -10,7 +10,7 @@ jobs:
 - job: Linux
 ###########################################
   variables:
-    COMPILER: clang
+    COMPILER: gcc
   pool:
     vmImage: 'ubuntu-16.04'
   container: ubuntu1404

--- a/.vsts-ci.yml
+++ b/.vsts-ci.yml
@@ -4,7 +4,7 @@ variables:
 resources:
   containers:
   - container: ubuntu1404
-    image: lightgbm/vsts-agent:ubuntu-14.04
+    image: lightgbm/vsts-agent:ubuntu-14.04-dev
 jobs:
 ###########################################
 - job: Linux
@@ -15,7 +15,7 @@ jobs:
     vmImage: 'ubuntu-16.04'
   container: ubuntu1404
   strategy:
-    maxParallel: 6
+    maxParallel: 7
     matrix:
       regular:
         TASK: regular
@@ -25,6 +25,8 @@ jobs:
       bdist:
         TASK: bdist
         PYTHON_VERSION: 3.6
+      swig:
+        TASK: swig
       inference:
         TASK: if-else
       mpi_source:

--- a/.vsts-ci.yml
+++ b/.vsts-ci.yml
@@ -4,7 +4,7 @@ variables:
 resources:
   containers:
   - container: ubuntu1404
-    image: lightgbm/vsts-agent:ubuntu-14.04-dev
+    image: lightgbm/vsts-agent:ubuntu-14.04
 jobs:
 ###########################################
 - job: Linux

